hugo/content/installation/configuration.md
---
title: "Configuration"
date:
draft: false
weight: 400
---

`pg_featureserv` provides a number of configuration options to control its operation.
Configuration can be provided in a number of ways:

* via a configuration file
* by environment variables
* on the command-line

## Configuration file

The configuration file is automatically read from from the following locations, if it exists:

* In the system configuration directory, at `/etc/pg_featureserv.toml`
* Relative to the directory from which the program is run, `./config/pg_featureserv.toml`
* In a root volume at `/config/pg_featureserv.toml`

If you want to specify a different file, use the `--config` commandline parameter to pass in a full path to the configuration file.  When using the `--config` option, the local configuration file is ignored.

```sh
./pg_featureserv --config /opt/pg_featureserv/config.toml
```

If no configuration is specified, the server runs using internal defaults
(which are the same as provided in the example configuration file below).
Where possible, the program autodetects values such as the `UrlBase`.

The only required configuration is the `DbConnection` setting,
if not provided in the environment variable `DATABASE_URL`.
(This is not required if the server is run with the `--test` flag.)

### Configuration Using Environment Variables

To set the database connection the environment variable `DATABASE_URL`
can be used with a
Postgres [connection string](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING):
```bash
export DATABASE_URL="host=localhost user=postgres"
```

Other parameters in the configuration file can be over-ridden in the environment.
Prepend the upper-cased parameter name with `PGFS_section_` to set the value.
For example, to change the HTTP port, function schemas, and service title:
```bash
export PGFS_SERVER_HTTPPORT=8889
export PGFS_DATABASE_FUNCTIONINCLUDES="postgisftw,my_funs"
export PGFS_METADATA_TITLE="My PGFS"
```

### Example Configuration

An example configuration file is shown below.

```toml
[Server]
# Accept connections on this subnet (default accepts on all)
HttpHost = "0.0.0.0"

# Accept connections on this port
HttpPort = 9000
#HttpsPort = 9001

# HTTPS configuration - TLS server certificate full chain and key
# If these are not specified, the TLS server will not be started
#TlsServerCertificateFile = "cert.pem"
#TlsServerPrivateKeyFile = "key.pem"

# Advertise URLs relative to this server name and path
# default is to look this up from incoming request headers
# Note: do not add a trailing slash.
# UrlBase = "http://localhost:9000/"

# Optional path to add to the service base URL
# If set, all routes will be prefixed with this path
# (e.g. "/pg_featureserv", "/services/pg_featureserv", etc.)
# Note: No trailing slash is necessary, pg_featureserv automatically
# adds a trailing slash for you.
# BasePath = "/"

# String to return for Access-Control-Allow-Origin header
# CORSOrigins = "*"

# set Debug to true to run in debug mode (can also be set on cmd-line)
# Debug = true

# Read html templates from this directory
AssetsPath = "/usr/share/pg_featurserv/assets"

# Maximum duration for reading entire request (in seconds)
ReadTimeoutSec = 1

# Maximum duration for writing response (in seconds)
# Also controls maximum time for processing request
WriteTimeoutSec = 30

[Database]
# Database connection
# postgresql://username:password@host/dbname
# DbConnection = "postgresql://username:password@host/dbname"

# Close pooled connections after this interval
# 1d, 1h, 1m, 1s, see https://golang.org/pkg/time/#ParseDuration
# DbPoolMaxConnLifeTime = "1h"

# Hold no more than this number of connections in the database pool
# DbPoolMaxConns = 4

# Publish only these schemas and tables (default is to publish all spatial tables)
# TableIncludes = [ "public", "priv_schema.tbl" ]

# Do not publish these schemas and tables
# TableExcludes = [ "priv_schema", "public.my_tbl" ]

# Publish functions from these schemas (default is publish postgisftw)
# FunctionIncludes = [ "postgisftw", "schema2" ]

# Designate a column as the feature ID (where primary key is not available e.g. views/functions)
# IDColumn = "id"

[Paging]
# The default number of features in a response
LimitDefault = 20
# Maxium number of features in a response
LimitMax = 10000

[Metadata]
# Title for this service
#Title = "pg-featureserv"
# Description of this service
#Description = "Crunchy Data Feature Server for PostGIS"

[Website]
# URL for the map view basemap
BasemapUrl = "https://maps.wikimedia.org/osm-intl/{z}/{x}/{y}.png"
```

### Configuration options

#### HttpHost

The IP address at which connections are accepted.

#### HttpPort and HttpsPort

The IP ports at which connections are accepted.

#### HTTPS configuration

If the `TlsServerCertificateFile` and `TlsServerPrivateKeyFile`
are specified then HTTPS support will be enabled,
at the port specified by `HttpsPort`.

#### UrlBase

The Base URL is the URL endpoint at which the service is advertised.
It is also used for any URL paths published by the service
(such as URLs for links in response documents).

The `UrlBase` parameter specifies a value for the Base URL. This accomodates running the service behind a reverse proxy.

The provided URL should not have a trailing slash.

##### Example
```
UrlBase = https://my-server.org/features
```

If `UrlBase` is not set, `pg_featureserv` dynamically detects the base URL.
Also, if the HTTP headers `Forwarded` or `X-Forwarded-Proto` and `X-Forwarded-Host` are present, they are respected.
Otherwise the base URL is determined by inspecting the incoming request.

#### BasePath

The BasePath allows an user to change the endpoint where pg_featureserv is served from.
If `BasePath` is not set, all traffic is served from the root of the website `localhost:9000/`.

##### Example
```
BasePath = "/services/pg_featureserv"
```

All traffic would then be served from `localhost:9000/services/pg_featurserv`.

#### CORSOrigins

The string to return in the `Access-Control-Allow-Origin` HTTP header,
which allows providing **Cross-Origin Resource Sharing** (CORS).

#### Debug

Set to `true` to run in debug mode.  This provides debug-level logging.

#### AssetsPath

The directory containing file assets used by the service (such as the HTML templates). It may be more convenient to deploy the asset files
in a location which is not relative to the service application path.

#### ReadTimeoutSec

The maximum duration (in seconds) the service allows for reading the HTTP request.
This can be relatively short, since service requests are small.

#### WriteTimeoutSec

The maximum duration (in seconds) the service allows for
processing and writing the HTTP response.
This should be long enough to allow expected requests to complete,
but not so long that the service can be saturated
by long-running requests.
Long request times may be caused by long execution times for database queries or functions,
or by returning very large responses.

#### DbConnection

The connection to the database can be set in this parameter,
using a Postgres [connection string](https://www.postgresql.org/docs/12/libpq-connect.html#LIBPQ-CONNSTRING).
The database connection can also be set via the `DATABASE_URL` environment variable, which takes precedence over this parameter.

#### DbPoolMaxConnLifeTime

The maximum duration for the lifetime for a pooled connection.
Specified using a Go [duration constant](https://golang.org/pkg/time/#ParseDuration)
such as `1d`, `2.5h`, or `30m`.

#### DbPoolMaxConns

The maximum number of database connections held in the connection pool.

#### TableIncludes

A list of the schemas and tables to publish feature collections from.
The default is to publish all geometry tables.

#### TableExcludes

A list of schemas and tables not to publish.
Overrides items specified in `TableIncludes`.

#### FunctionIncludes

A list of the schemas to publish functions from.
The default is to publish functions in the `postgisftw` schema.

#### IDColumn

The column to use as the feature ID in cases where a primary key is not available. The default is `id`.

> NOTE: The presence of a primary key will supercede this setting.

#### LimitDefault

The default number of features in a response,
if not specified by the `limit` [query parameter](/usage/query_data/).

#### LimitMax

The maximum number of features that can be returned in a response.
This cannot be overridden by the `limit` query paramater.

#### Title

The title for the service.
Appears in the HTML web pages, JSON responses, and the log.

#### Description

The description for the service.
Appears in the HTML web pages and JSON responses.

#### BasemapUrl

The URL template for the basemap used in the web UI map views.
Must be a URL template suitable for the OpenLayers OSM class.


NEWS.md
# pg_featureserv Version History

## Version ??
*Released: TBD*

### Bug Fixes

* Fix CQL parser to allow multiple AND/OR terms (#162)
* Fix collection JSON output to use nested extent bbox as per spec (#175)

## Version 1.3.1
*Released: 2024 Feb 6*

### Improvements

* Container build update (#154)
* Improve map UI layout (#146)

### Bug Fixes

* Fix CQL parser to allow scientific notation using `e` (#141)


## Version 1.3
*Released: 2023 July 10*

### New Features

* Support for HTTPS
* Support Postgres data types: JSON, array of text/int/float/numeric
* Add support for partitioned tables and foreign tables
* Add support for tables storing `geography`
* Add configuration to include/exclude published schemas and tables
* Add configuration for web UI map view basemap URL template
* Add configuration to set base path (#88)
* Add configuration via enviroment variables (#104)
* Allow configuring published function schemas (#99)
* Add the `groupby` query parameter, and ability to aggregate query features
* Support OGC API query parameter `sortby` (`orderBy` is deprecated)
* Add OGC API query parameters `crs` and `bbox-crs` (#98)
* Add CQL filtering query parameters `filter` and `filter-crs`  (#101, #102, #103, #105)

### Performance Improvements

* Improve performance of page template loading using caching

### Bug Fixes

* Fix response value for empty query results for collections and functions
* Fix OpenAPI document to respect configuration `UrlBase`
* Quote query table and function names to support mixed case
* Fix `collection` reponse to include actual extent (#54)
* Fix `collection` response JSON `extent` property to be OAPIF-compliant (#54)
* Fix landing page document to include links to resources `service-desc` and `conformance`
* Fix links in JSON responses to not include JSON format (required for QGIS)
* Fix function feature requests to respect the `offset` parameter (#65)
* Fix Swagger request to work under HTTPS (#71)
* Fix Collection description to be populated from table description (#70)
* Fix output of NULL geometry values to be JSON null
* Fix `offset` parameter to allow any non-negative value
* Fix encoding of primary key column names in item requests (#80)
* Fix error handling during Feature reading from database (#96)
* Check for empty strings in Forwarded header parsing

## Version 1.2
*Released: 2020 Dec 1*

### New Features

* Display materialized views as collections
* Improve logging and error messages

### Bug Fixes

* Handle specialty column types by casting to text (#52)

## Version 1.1
*Released: 2020 Apr 6*

### New Features

* Query parameters for filtering by property value
* Configuration write timeout cancels service processing (but not DB process)
* Handle client cancellation
* Improved collection metadata JSON doc
* Data-driven styling for points
* Data-driven styling column `style_radius`
* Configuration whitelist for geometry transformation functions
* Quote database columns in queries to allow upper case, reserved SQL words
* More detailed debug logging


### Bug Fixes

* Fix handling Accept headers to work for all paths
* Fix configuration loading to give priority to env var DATABASE_URL over config file
* Fix function argument ordering in database metadata query
* Fix function metadata query to only show functions with `execute` privilege
* Fix handling of functions with no default parameters
* Fix Home page JSON link

## Version 1.0.0
*Released: 2020 Feb 29*


internal/conf/config.go
package conf

/*
 Copyright 2019 - 2025 Crunchy Data Solutions, Inc.
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

import (
	"fmt"
	"os"
	"strings"

	log "github.com/sirupsen/logrus"
	"github.com/spf13/viper"
)

// Configuration for system
var Configuration Config

func setDefaultConfig() {
	viper.SetDefault("Server.HttpHost", "0.0.0.0")
	viper.SetDefault("Server.HttpPort", 9000)
	viper.SetDefault("Server.HttpsPort", 9001)
	viper.SetDefault("Server.TlsServerCertificateFile", "")
	viper.SetDefault("Server.TlsServerPrivateKeyFile", "")
	viper.SetDefault("Server.UrlBase", "")
	viper.SetDefault("Server.BasePath", "")
	viper.SetDefault("Server.CORSOrigins", "*")
	viper.SetDefault("Server.Debug", false)
	viper.SetDefault("Server.AssetsPath", "./assets")
	viper.SetDefault("Server.ReadTimeoutSec", 5)
	viper.SetDefault("Server.WriteTimeoutSec", 30)

	viper.SetDefault("Database.DbPoolMaxConnLifeTime", "1h")
	viper.SetDefault("Database.DbPoolMaxConns", 4)
	viper.SetDefault("Database.TableIncludes", []string{})
	viper.SetDefault("Database.TableExcludes", []string{})
	viper.SetDefault("Database.FunctionIncludes", []string{"postgisftw"})
	viper.SetDefault("Database.IDColumn", "id")

	viper.SetDefault("Paging.LimitDefault", 10)
	viper.SetDefault("Paging.LimitMax", 1000)

	viper.SetDefault("Metadata.Title", "pg-featureserv")
	viper.SetDefault("Metadata.Description", "Crunchy Data Feature Server for PostGIS")

	viper.SetDefault("Website.BasemapUrl", "https://{a-d}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}.png")
}

// Config for system
type Config struct {
	Server   Server
	Paging   Paging
	Metadata Metadata
	Database Database
	Website  Website
}

// Server config
type Server struct {
	HttpHost                 string
	HttpPort                 int
	HttpsPort                int
	TlsServerCertificateFile string
	TlsServerPrivateKeyFile  string
	UrlBase                  string
	BasePath                 string
	CORSOrigins              string
	Debug                    bool
	AssetsPath               string
	ReadTimeoutSec           int
	WriteTimeoutSec          int
	TransformFunctions       []string
}

// Paging config
type Paging struct {
	LimitDefault int
	LimitMax     int
}

// Database config
type Database struct {
	DbConnection          string
	DbPoolMaxConnLifeTime string
	DbPoolMaxConns        int
	TableIncludes         []string
	TableExcludes         []string
	FunctionIncludes      []string
	IDColumn              string
}

// Metadata config
type Metadata struct {
	Title       string //`mapstructure:"METADATA_TITLE"`
	Description string
}

type Website struct {
	BasemapUrl string
}

// IsHTTPSEnabled tests whether HTTPS is enabled
func (conf *Config) IsTLSEnabled() bool {
	return conf.Server.TlsServerCertificateFile != "" && conf.Server.TlsServerPrivateKeyFile != ""
}

// InitConfig initializes the configuration from the config file
func InitConfig(configFilename string, isDebug bool) {
	// --- defaults
	setDefaultConfig()

	if isDebug {
		viper.Set("Debug", true)
	}

	viper.SetEnvPrefix(AppConfig.EnvPrefix)
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv()

	isExplictConfigFile := configFilename != ""
	confFile := AppConfig.Name + ".toml"
	if configFilename != "" {
		viper.SetConfigFile(configFilename)
	} else {
		viper.SetConfigName(confFile)
		viper.SetConfigType("toml")
		viper.AddConfigPath("./config")
		viper.AddConfigPath("/config")
		viper.AddConfigPath("/etc")
	}
	err := viper.ReadInConfig() // Find and read the config file
	if err != nil {
		_, isConfigFileNotFound := err.(viper.ConfigFileNotFoundError)
		errrConfRead := fmt.Errorf("fatal error reading config file: %s", err)
		isUseDefaultConfig := isConfigFileNotFound && !isExplictConfigFile
		if isUseDefaultConfig {
			log.Debug(errrConfRead)
		} else {
			log.Fatal(errrConfRead)
		}
	}

	log.Infof("Using config file: %s", viper.ConfigFileUsed())
	errUnM := viper.Unmarshal(&Configuration)
	if errUnM != nil {
		log.Fatal(fmt.Errorf("fatal error decoding config file: %v", errUnM))
	}

	// Read environment variable database configuration
	// It takes precedence over config file (if any)
	// A blank value is ignored
	dbconnSrc := "config file"
	if dbURL := os.Getenv(AppConfig.EnvDBURL); dbURL != "" {
		Configuration.Database.DbConnection = dbURL
		dbconnSrc = "environment variable " + AppConfig.EnvDBURL
	}
	log.Infof("Using database connection info from %v", dbconnSrc)

	// sanitize the configuration
	Configuration.Server.BasePath = strings.TrimRight(Configuration.Server.BasePath, "/")
}

func DumpConfig() {
	log.Debugf("--- Configuration ---")
	//fmt.Printf("Viper: %v\n", viper.AllSettings())
	//fmt.Printf("Config: %v\n", Configuration)
	var basemapURL = Configuration.Website.BasemapUrl
	if basemapURL == "" {
		basemapURL = "*** NO URL PROVIDED ***"
	}
	log.Debugf("  BasemapUrl = %v", basemapURL)
	log.Debugf("  TableIncludes = %v", Configuration.Database.TableIncludes)
	log.Debugf("  TableExcludes = %v", Configuration.Database.TableExcludes)
	log.Debugf("  FunctionIncludes = %v", Configuration.Database.FunctionIncludes)
	log.Debugf("  TransformFunctions = %v", Configuration.Server.TransformFunctions)
}
